testnet/stacks-node/src/nakamoto_node/relayer.rs

```diff
@@ -852,104 +877,219 @@ impl RelayerThread {
+            (
+                StacksBlockId(last_snapshot.winning_stacks_block_hash.0),
+                last_snapshot,
+                MinerReason::EmptyTenure,
+            )
+        } else {
+            debug!("Relayer: Successfully issued a tenure change payload. Issue a continue extend from the chain tip.");
+            (
+                self.get_canonical_stacks_tip(),
+                canonical_snapshot,
+                MinerReason::Extended {
+                    burn_view_consensus_hash: new_burn_view,
+                },
+            )
         }
-        debug!("Relayer: successfully stopped tenure.");
-        // Check if we should undergo a tenure change to switch to the new burn view
-        let burn_tip =
-            SortitionDB::get_block_snapshot_consensus(self.sortdb.conn(), &new_burn_view)
+    }
+
+    /// Get the block snapshot of the most recent sortition that committed to
+    /// the canonical tip. If the latest sortition did not commit to the
+    /// canonical tip, then the tip's tenure is the last good sortition.
+    fn get_last_good_block_snapshot(
+        &self,
+        burn_tip: &BlockSnapshot,
+        highest_tenure_bhh: &BlockHeaderHash,
+        canonical_stacks_tip_ch: &ConsensusHash,
+    ) -> Result<BlockSnapshot, NakamotoNodeError> {
+        let ih = self.sortdb.index_handle(&burn_tip.sortition_id);
+        let sn = ih
+            .get_last_snapshot_with_sortition(burn_tip.block_height)
+            .map_err(|e| {
+                error!("Relayer: failed to get last snapshot with sortition: {e:?}");
+                NakamotoNodeError::SnapshotNotFoundForChainTip
+            })?;
+        if &sn.winning_stacks_block_hash != highest_tenure_bhh {
+            info!(
+                "Relayer: Sortition winner is not committed to the canonical tip; allowing last miner to extend";
+                "burn_block_height" => burn_tip.block_height,
+                "consensus_hash" => %burn_tip.consensus_hash,
+            );
+
+            SortitionDB::get_block_snapshot_consensus(self.sortdb.conn(), canonical_stacks_tip_ch)
                 .map_err(|e| {
-                    error!("Relayer: failed to get block snapshot for new burn view: {e:?}");
+                    error!("Relayer: failed to get block snapshot for canonical tip: {e:?}");
                     NakamotoNodeError::SnapshotNotFoundForChainTip
                 })?
                 .ok_or_else(|| {
-                    error!("Relayer: failed to get block snapshot for new burn view");
+                    error!("Relayer: failed to get block snapshot for canonical tip");
                     NakamotoNodeError::SnapshotNotFoundForChainTip
-                })?;
+                })
+        } else {
+            Ok(sn)
+        }
+    }
+
+    fn get_block_snapshot(&self, ch: &ConsensusHash) -> Result<BlockSnapshot, NakamotoNodeError> {
```

Comment:
It looks like you're wrapping a lot of existing methods in `SortitionDB` and `NakamotoChainState` with private functions here, where the only real differences are in that you're logging errors and converting the errors into `NakamotoNodeError` variants.  It may be more economical to instead implement `From<ChainstateError>` for `NakamotoNodeError`, and add a text field to `NakamotoNodeError` to include the `error!()` text.

Reply:
Yeah, the wrapper functions were part of my effort to try to clean up `continue_tenure` because it was getting really big and confusing. I will give this a try to see if it helps.